File: Qt e Progetti/Progetti/2017-2018 Progetto Calcolatrice/Polygonator - Menon e Dalla Via/Java/src/Model/Polygons/Rectangle.java
package Model.Polygons;

import Model.Line;
import Model.Point;

import java.util.List;

public class Rectangle extends Parallelogram {
    public Rectangle(final List<Point> vertices) throws Exception {
        super(vertices);
    }

    /**
     * Returns whether a polygon is a rectangle or not
     */
    public static boolean canBeCreatedWith(final Polygon polygon) throws Exception {
        if (!Parallelogram.canBeCreatedWith(polygon)) return false;

        for (int i = 0; i < polygon.getVertices().size(); i++) {
            Point p = i == 0 ? polygon.getVertices().get(polygon.getVertices().size() - 1) : polygon.getVertices().get(i - 1);
            Point n = i == polygon.getVertices().size() - 1 ? polygon.getVertices().get(0) : polygon.getVertices().get(i + 1);
            Point a = polygon.getVertices().get(i);

            if (Math.abs(Line.getAngleDegrees(p, a, n) - 90) > 0.0001) return false;
        }

        return true;
    }

    @Override
    public double getHeight() {
        return Line.getDistance(vertices.get(0), vertices.get(vertices.size() - 1));
    }

    @Override
    public double getPerimeter() throws Exception {
        return (getBase() + getHeight()) * 2;
    }

    @Override
    public String toString() {
        return "Rectangle";
    }
}

File: Qt e Progetti/Progetti/2017-2018 Progetto Calcolatrice/Polygonator - Menon e Dalla Via/C++/Model/Polygons/Rectangle.cpp
#include "Rectangle.h"

Rectangle::Rectangle(const list<Point> &vertices) : Quadrilateral(vertices), Parallelogram(vertices) {}

bool Rectangle::canBeCreatedWith(const Polygon &polygon) {
    if (!Parallelogram::canBeCreatedWith(polygon)) return false;

    for (auto it = polygon.getVertices().begin(); it != polygon.getVertices().end(); it++) {
        Point p = it == polygon.getVertices().begin() ? *prev(polygon.getVertices().end()) : *prev(it);
        Point n = it == prev(polygon.getVertices().end()) ? *polygon.getVertices().begin() : *next(it);

        if (fabs(Line::getAngleDegrees(p, *it, n) - 90) > 0.0001) return false;
    }

    return true;
}

double Rectangle::getHeight() const {
    return Line::getDistance(*prev(getVertices().end()), *getVertices().begin());
}

double Rectangle::getPerimeter() const {
    return (getBase() + getHeight()) * 2;
}

string Rectangle::getName() const {
    return "Rectangle";
}
